### ISSUE TITLE
Patch for error bars on points

### ISSUE THREAD

**Issue by GoogleCodeExporter** [josejaner/flot#215]
```
Here is a patch for the current flot SVN (198) that adds error bar plotting 
capabilities, essential for most scientific data visualizations.

The patch allows you to define X and/or Y (no X2 or Y2 support for now) error 
bars on any series of points. They may be symmetric or asymmetric (different 
upper and lower errors) and the drawing function of the lower/upper caps of 
the bars may be defined by the user. Axis can also be set to automatically 
scale taking the error bars into account.

Refer to the attached HTML example for all the different possibilites. 
Comments/suggestions are welcome.
```

Original issue reported on code.google.com by `rui.pereira` on 30 Aug 2009 at 9:51

**Comment by GoogleCodeExporter** [comment]
```
Issue 196 has been merged into this issue.
```

Original comment by `olau%iol...@gtempaccount.com` on 7 Oct 2009 at 4:25

**Comment by GoogleCodeExporter** [comment]
```
Hi!

I'm going to try it, it sounds neat. However, have you considered doing it as a
plugin? I'm trying to move things out of the core right now, to make it easier 
for
people to customize them without having to do a lot of merge work each time a 
new
version of Flot is out.
```

Original comment by `olau%iol...@gtempaccount.com` on 7 Oct 2009 at 4:30
- Added labels: **Type-Enhancement**
- Removed labels: **Type-Defect**

**Comment by GoogleCodeExporter** [comment]
```
Hello. Yes, I agree a plugin would be cleaner, I'm gonna take a look into that 
soon (I hope). Cheers!
```

Original comment by `rui.pereira` on 7 Oct 2009 at 4:55

**Comment by GoogleCodeExporter** [comment]
```
Hi Team-iola, 
We see your post in regards to Error bars plotting , can you share with us your
experience after trying out the latter patch proivided by Rui. 
As of now looking out a way around for this Erro bar on a Line Graph. 
Any help is appreciated.

Rajesh.

```

Original comment by `rajesh.h...@gmail.com` on 12 Oct 2009 at 7:54

**Comment by GoogleCodeExporter** [comment]
```
Rajest: Sorry, haven't tried it yet, still got a backlog of other things to do.
```

Original comment by `olau%iol...@gtempaccount.com` on 22 Oct 2009 at 12:11

**Comment by GoogleCodeExporter** [comment]
```
Hi. Here's a plugin compatible with flot 0.6 that adds error bars capabilities 
to the points series. It has the 
same functionality as the patch I submitted before (X,Y,XY error bars, 
asymmetric errors, user defined caps) 
but it's more modular and avoids messy patching.

For Rajesh, you can also use it to draw error bars on a line graph, as long as 
you use (radius: 0) points 
connected with lines.

See the attached HTML for application examples.

Rui
```

Original comment by `rui.pereira` on 26 Oct 2009 at 11:54

**Comment by GoogleCodeExporter** [comment]
```
Tiny bug fix, it was working OK in Safari but not under Firefox or Chrome.
```

Original comment by `rui.pereira` on 27 Oct 2009 at 11:54

**Comment by GoogleCodeExporter** [comment]
```
Error bars is the most important feature after plotting the data points itself. 
What meaning the data point have to anybody if one does not know how sure the 
value is?

Why developers won't merge this patch into the main steam release instead of 
refereeing here?

Flot would than be the only javascript plotting framework on the world to be 
usable for the scientific community. For now there is none.

```

Original comment by `jan.skow...@gmail.com` on 4 Sep 2010 at 9:33

**Comment by GoogleCodeExporter** [comment]
```
I agree that error bars should be in the main release because in a lot of cases 
no error bars means the package can't be used. 

Thanks very much Rui for posting the package! I also wanted to make some very 
small modifications to the above posted code. With the developer flot releases 
you can pan your plot by clicking and dragging and can also easily zoom, the 
errorbars package Rui posted has a few problems when doing this however, as the 
error bars can be drawn outside of the plot canvas, especially when zooming and 
panning. For a standard plot with left and bottom tick marks adding a few lines 
cleans this problem up:

to line 195 something like:
           //if the point is outside of x plotting range break from plotting loop
                    if (x < ax[0].min || x > ax[0].max)
                    {
                        break;
                    }
to line 217 something like this:
                     //Prevent the upper error bar from being drawn below the bottom axis
                     if (y <= ax[e].min) {
                         y = ax[e].min;
                     }

These additions also significantly smooth the panning and zooming when you are 
looking at a small portion of a larger data set.
```

Original comment by `AdamBSim...@gmail.com` on 15 Sep 2010 at 5:07

**Comment by GoogleCodeExporter** [comment]
```
Hi Adam,

Could you show an example where the errorbars plugin isn't working while 
panning and zooming? The changes you propose will prevent any point that is out 
of the canvas (in X) to be drawn, even if its errorbar is partially visible 
(eg. x = 2 ± 1 with xmax = 1.5). I tried to make the plugin aware of those 
cases, and for me it works even with panning/zooming. Please check the attached 
example.
```

Original comment by `rui.pereira` on 16 Sep 2010 at 8:31

**Comment by GoogleCodeExporter** [comment]
```
Thanks for taking a look at it Rui,

I have included an image of the problem I am experiencing with the file you 
posted, you can see that error bars are going outside of the plot region where 
the axis numbers are located. The main problem is when you have say something 
like { x = 2 , y = 2 ± 1} with xmax = 1.5, I do not believe the code checks 
that the error bars in y have an x position that is inside of the plot region, 
I believe it only checks that error bars in y have a y position that is inside 
of the plot area. This is for example shown by the green vertical error bar 
centered on the y = 2.0 marker in the image attached.

I also believe that if you have error bars in y and say are dragging the plot 
downwards it will only shorten the length of the error bar below the point, 
drawing from the points y location to the bottom of the chart that is showing. 
Once you have dragged the plot such that only error bar above the point is 
showing the code does not fix the length correctly. This for example is shown, 
although just slightly, in the red vertical error bar centered on the x = 1.5 
marker in the image.

The changes I made certainly are not complete solutions or particularly well 
thought out but seemed to fix more or less the problems with vertical error 
bars as seen in the attached image. Thanks again for the package, without it I 
don't know what I would do!
```

Original comment by `AdamBSim...@gmail.com` on 17 Sep 2010 at 12:38

Attachments:
- [flot-problem.png](https://storage.googleapis.com/google-code-attachments/flot/issue-215/comment-11/flot-problem.png)

**Comment by GoogleCodeExporter** [comment]
```
Oh, dumb me, I had already fixed this bug on my local plugin, but never got to 
upload it here. So that's why I was not able to reproduce your problem...

Anyway, here's the most recent version of the errorbars plugin, hope it works 
for you.
```

Original comment by `rui.pereira` on 17 Sep 2010 at 8:09

Attachments:
- [jquery.flot.errorbars.js](https://storage.googleapis.com/google-code-attachments/flot/issue-215/comment-12/jquery.flot.errorbars.js)
- [errorbars-example.html](https://storage.googleapis.com/google-code-attachments/flot/issue-215/comment-12/errorbars-example.html)

**Comment by GoogleCodeExporter** [comment]
```
Can someone please send a pull request for this change on 
https://github.com/flot/flot ?  I'm using it and it's working great, thanks for 
writing it!
```

Original comment by `tsuna...@gmail.com` on 7 Mar 2012 at 11:54

**Comment by GoogleCodeExporter** [comment]
```
Pull request issued: https://github.com/flot/flot/pull/41
```

Original comment by `rui.pereira` on 9 Mar 2012 at 10:05

**Comment by GoogleCodeExporter** [comment]
Original comment by `dnsch...@gmail.com` on 7 May 2012 at 11:19
- Changed state: **Accepted**

**Comment by GoogleCodeExporter** [comment]
```
Resolved by merging pull request #41.
```

Original comment by `dnsch...@gmail.com` on 16 Jul 2012 at 4:23
- Changed state: **Verified**
